backend/groth16/bls12-377/prove.go

```diff
@@ -61,35 +63,32 @@ func Prove(r1cs *cs.R1CS, pk *ProvingKey, fullWitness witness.Witness, opts ...b
 
 	log := logger.Logger().With().Str("curve", r1cs.CurveID().String()).Int("nbConstraints", r1cs.GetNbConstraints()).Str("backend", "groth16").Logger()
 
-	proof := &Proof{}
+	proof := &Proof{Commitments: make([]curve.G1Affine, len(r1cs.CommitmentInfo))}
 
 	solverOpts := opt.SolverOpts[:len(opt.SolverOpts):len(opt.SolverOpts)]
 
-	if r1cs.GetNbCommitments() != 0 {
-		solverOpts = append(solverOpts, solver.OverrideHint(r1cs.CommitmentInfo[0].HintID, func(_ *big.Int, in []*big.Int, out []*big.Int) error {
-			// Perf-TODO: Converting these values to big.Int and back may be a performance bottleneck.
-			// If that is the case, figure out a way to feed the solution vector into this function
-			if len(in) != r1cs.CommitmentInfo[0].NbCommitted() { // TODO: Remove
-				return fmt.Errorf("unexpected number of committed variables")
-			}
-			values := make([]fr.Element, r1cs.CommitmentInfo[0].NbPrivateCommitted)
-			nbPublicCommitted := len(in) - len(values)
-			inPrivate := in[nbPublicCommitted:]
-			for i, inI := range inPrivate {
-				values[i].SetBigInt(inI)
-			}
+	commitmentSubIndexes := r1cs.CommitmentInfo.CommitmentIndexesInCommittedLists()
+	privateCommittedValues := make([][]fr.Element, len(r1cs.CommitmentInfo))
+	for i := range r1cs.CommitmentInfo {
+		solverOpts = append(solverOpts, solver.OverrideHint(r1cs.CommitmentInfo[i].HintID, func(i int) solver.Hint {
+			return func(_ *big.Int, in []*big.Int, out []*big.Int) error {
+				privateCommittedValues[i] = make([]fr.Element, r1cs.CommitmentInfo[i].NbPrivateCommitted-len(commitmentSubIndexes[i]))
+				hashed, committed := internal.DivideByThresholdOrList(r1cs.CommitmentInfo[i].NbPublicCommitted(), commitmentSubIndexes[i], in)
+				for j, inJ := range committed {
+					privateCommittedValues[i][j].SetBigInt(inJ) // TODO @Tabaie Perf If this takes significant time can read values off the witness vector instead
```

Comment:
`SetBigInt` is probably okay overhead vs `Set` (since `in` is already allocating `big.Int`). 1 extra mul per value.

However, I think `DivideByThresholdOrList` should work in-place, and re-slice the `in` input slice into 2. Avoid allocating too many slices of pointers to `big.Int` will matter.